### DISCUSSION TITLE
[WIP] Communication protocol under centralized coordination

### DISCUSSION THREAD

**Discussion by lsk567** [lf-lang/lingua-franca#1385] [Specifications]
This is a work-in-progress. All contributions and comments are welcome.

## Abstract
This document specifies the communication protocol used in the federated execution of Lingua Franca under centralized coordination, which consists of a set of federates (top-level reactors each with a network interface) and a runtime infrastructure (RTI). The federates exchange standardized messages with the RTI at runtime to advance their current logical tags.

There have been many contributors to this work both in terms of concepts, code, and text. This document aims to unify the comments in the existing implementation and concepts presented in previous papers and to present a concise and visual presentation of this protocol to help drive the implementation of centralized coordination in other runtimes and contexts (such as "affiliation").

### Message Types
Types |	ID | Payload (prefixed by # bytes) | Description
:---: | :---: | --- | ---
REJECT | 0 | n/a | Byte identifying a rejection of the previously received message. The reason for the rejection is included as an additional byte (uchar) (see below for encodings of rejection reasons).
ACK | 255 | n/a | Byte identifying an acknowledgment of the previously received message. This message carries no payload.
UDP_PORT | 254 | [1] UDP port for clock sync. <br/> [4] UDP server port | Byte identifying an acknowledgment of the previously received FED_IDS message sent by the RTI to the federate with a payload indicating the UDP port to use for clock synchronization. The next four bytes will be the port number for the UDP server, or 0 or USHRT_MAX if there is no UDP server. 0 means that initial clock synchronization is enabled, whereas USHRT_MAX mean that no synchronization should be performed at all.
FED_IDS | 1 | n/a | n/a
| Stop request         | 10   | [8] Timestamp at which the federate want to stop <br />[4] Microstep at which the federate want to stop |Byte identifying a stop request. This message is first sent to the RTI by a federate that would like to stop execution at the specified tag. The RTI will forward the MSG_TYPE_STOP_REQUEST to all other federates. Those federates will either agree to the requested tag or propose a larger tag. The RTI will collect all proposed tags and broadcast the largest of those to all federates. All federates will then be expected to stop at the granted tag.|
| Stop request reply   | 11   | [8] Timestamp at which the federate can be stopped <br />[4] Microstep at which the federate can be stopped | Byte indicating a federate's reply to a MSG_TYPE_STOP_REQUEST that was sent by the RTI. The payload is a proposed stop tag that is at least as large as the one sent to the federate in a MSG_TYPE_STOP_REQUEST message.|
| Stop request granted | 12   | [8] Timestamp at which the federates will stop<br /> [4] Microstep at which the federates will stop | Byte sent by the RTI indicating that the stop request from some federate has been granted. The payload is the tag at which all federates have agreed that they can stop.

## Rationale
FIXME

## Reference Implementation
- https://github.com/lf-lang/reactor-c/blob/main/core/federated/federate.c
- https://github.com/lf-lang/reactor-c/blob/main/core/federated/RTI/rti.c

## Notes
This document is styled after Ethereum's EIP (Ethereum Improvement Proposal) and IETF's RFC (Request for Comment), which typically include a preamble on top. However, because Github Discussions cannot render preambles, this is temporarily moved here.
```
---
proposal id: <FIXME>
title: Lingua Franca communication protocol under centralized coordination
description: This document specifies the communication protocol used in the federated execution of Lingua Franca under centralized coordination. 
author: <FIXME>
status: Draft
created: <FIXME>
---
```
Here is the [template](https://github.com/ethereum/EIPs/blob/master/eip-template.md) for this document.

**Comment by hokeun** [comment]
@byeong-gil showed interest in contributing to the shutdown part (including the stop request) of this specification. @lsk567, is it ok if we directly edit the document above, maybe starting with the message types used for shutdown?

**Reply by lsk567** [reply]
@hokeun Sorry for the late reply. Absolutely! @byeong-gil can directly edit the document above. Thanks.

**Reply by byeonggiljun** [reply]
Hello, @lsk567. I made a table to describe how the shutdown messages work(see this [notion page](https://political-brush-428.notion.site/Shutdown-protocol-message-f40465b1c2de4eabb79ede59bfefb306)). Do you think it looks good to add this table to this document?

Additionally, what is a better way to affix this table to the original table, separate tables according to the role of message types(e.g. messages used in starting, shutdown, etc), or just append them?

**Reply by lsk567** [reply]
Hi @byeong-gil, apologies for the delayed response. Yes, the table describing the shutdown messages looks great to me. Thanks for working on it. For now, I think we can just append your table to the existing table. The message types should be obvious enough to show that the messages are used for shutdown.

**Comment by lhstrh** [comment]
How does this relate to https://github.com/lf-lang/website-lingua-franca/pull/104?

**Reply by lsk567** [reply]
After checking with @suyourice, it seems like we are doing similar things. The goal is to have a concise [RFC-style document](https://en.wikipedia.org/wiki/Request_for_Comments) that contains graphics illustrating the communication protocols. @suyourice and I agree to collaborate on this together. I am fine with hosting the doc either here or on the website.

**Comment by Rodario** [comment]
Is there a specific reason for associating with ethereum?
Its not used anywhere else in the project.

**Reply by lhstrh** [reply]
I don't think Shaokai's point is to associate with Ethereum. I think their EIP is just one of the more recent incarnations of RFC styles that is not purely text (ASCII) based.

**Reply by Rodario** [reply]
Got it.
